### PR TITLE
[NVIDIA XLA] Cover D_amax output fusion for FP8 GEMM when Relu is present in epilogue

### PR DESCRIPTION
**Pull request by wenscarl**
In the presence of a ReLU activation, algsimp pass simplifies the possible abs op following ReLU, like abs(Relu(x)) -> Relu(x). This PR covers that corner case by adding possible patterns to match.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 1]
```diff
@@ -901,16 +901,22 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     const std::vector<HloInstruction *> gemm_users = existing_gemm->users();
     HloInstruction *reduce_damax = nullptr;
     if (gemm_users.size() == 2) {
+      // Once Relu is fused, abs op will be elided since abs(Relu) = Relu.
```
Comment (philipphack) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:904]: Does this also apply when the ReLU is not fused?

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 2]
```diff
@@ -901,16 +901,22 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-        if (gemm_users[i]->opcode() == HloOpcode::kAbs &&
-            gemm_users[i]->users().size() == 1 &&
-            gemm_users[i]->users()[0]->opcode() == HloOpcode::kReduce &&
-            gemm_users[i]->users()[0]->operands().size() == 2 &&
-            gemm_users[i]->users()[0]->operand(1)->opcode() ==
-                HloOpcode::kConstant &&
-            ShapeUtil::IsScalar(
-                gemm_users[i]->users()[0]->operand(1)->shape())) {
-          HloInstruction *reduce = gemm_users[i]->users()[0];
+        HloInstruction *reduce =
```
Comment (philipphack) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:911]: At this point, it's not yet known whether this is actually a reduce instruction. I would pick a different name.

**Review by philipphack — CHANGES_REQUESTED** [review]

**Comment by cheshire** [comment]
@reedwm could you take a look?

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 3]
```diff
@@ -901,16 +901,24 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-            gemm_users[i]->users()[0]->operands().size() == 2 &&
-            gemm_users[i]->users()[0]->operand(1)->opcode() ==
-                HloOpcode::kConstant &&
-            ShapeUtil::IsScalar(
-                gemm_users[i]->users()[0]->operand(1)->shape())) {
-          HloInstruction *reduce = gemm_users[i]->users()[0];
+        HloInstruction *maybe_reduce =
+            has_relu_epilogue ? gemm_users[i] : gemm_users[i]->users()[0];
+
+        if ((has_relu_epilogue || gemm_users[i]->opcode() == HloOpcode::kAbs &&
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:915]: We should also allow the Abs to appear even with a Relu, unless a previous XLA pass removes the Abs for us by optimizing abs(relu(x)) to relu(x)

**Review by reedwm — CHANGES_REQUESTED** [review]
Please add a test.

Also maybe move the amax logic to a new function

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc (hunk 3 above)
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:915]: Yes. Algsimp pass removes it.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 4]
```diff
@@ -901,16 +901,24 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-            gemm_users[i]->users().size() == 1 &&
-            gemm_users[i]->users()[0]->opcode() == HloOpcode::kReduce &&
-            gemm_users[i]->users()[0]->operands().size() == 2 &&
-            gemm_users[i]->users()[0]->operand(1)->opcode() ==
-                HloOpcode::kConstant &&
-            ShapeUtil::IsScalar(
-                gemm_users[i]->users()[0]->operand(1)->shape())) {
-          HloInstruction *reduce = gemm_users[i]->users()[0];
+        HloInstruction *maybe_reduce =
+            has_relu_epilogue ? gemm_users[i] : gemm_users[i]->users()[0];
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:913]: Fixed.

**Comment by wenscarl** [comment]
> Please add a test.
> 
> Also maybe move the amax logic to a new function

The following "changing data type of C to BF16" part is also in the body of F8ConvertD.

File: tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc [hunk 5]
```diff
@@ -901,16 +901,27 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
...
-            gemm_users[i]->users()[0]->operand(1)->opcode() ==
-                HloOpcode::kConstant &&
-            ShapeUtil::IsScalar(
-                gemm_users[i]->users()[0]->operand(1)->shape())) {
-          HloInstruction *reduce = gemm_users[i]->users()[0];
+        if (!has_relu_epilogue && gemm_users[i]->users().empty()) {
+          continue;
+        }
+        HloInstruction *maybe_reduce =
+            has_relu_epilogue ? gemm_users[i] : gemm_users[i]->users()[0];
```
Comment (reedwm) [inline tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc:916]: This logic is somewhat hard to follow. I would replace these lines with something like

```c++
HloInstruction* maybe_reduce = nullptr;
if (gemm_users[i]->opcode() == HloOpcode::kAbs) {
  if (gemm_users[i]->users().size() != 1) continue;
  maybe_reduce = gemm_users[i]->users()[0];
} else {
  // If there is no Abs instruction, we require a relu epilogue to ensure all values are nonnegative.
  if (!has_relu_epilogue) continue;
  maybe_reduce = gemm_users[i];
}
```

This allows you to remove the `has_relu_epilogue` check and the `abs` check below as well. This logic also allows the Abs to appear with a Relu, which I didn't intend but doesn't hurt.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]
